httpcore5/src/main/java/org/apache/hc/core5/concurrent/BasicFuture.java

```diff
@@ -94,7 +95,7 @@ public synchronized T get(final long timeout, final TimeUnit unit)
         if (this.completed) {
             return getResult();
         } else if (waitTime <= 0) {
-            throw new TimeoutException();
+            throw TimeoutValueException.ofMillis(msecs, msecs + Math.abs(waitTime));
```

Comment:
@garydgregory What is the value of this information? What good does it make to print out current time and deadline in milliseconds? I thought you were going to include the timeout value in the exception message, which would be of some marginal use.

Reply:
@ok2c ,

Thank you for your comment.

What is the "timeout value"? Is it the boundary of time above which the timeout occurs or is it the actual elapsed time? To both avoid this confusion and to be as precise with the message as possible, both values are given: 

- The "deadline" in the message is the the boundary of time above which the timeout occurs.
- The "actual" in the message is the actual elapsed time.

Does that clarify things for you?

Gary

Reply:
@garydgregory To me it is Is it the _"boundary of time above which the timeout occurs"_. Yes, it does clarify but i personally do not find this information useful. The initial timeout value given by the user, say 30 seconds, might be marginally useful, but the number of milliseconds that have elapsed since midnight, January 1, 1970, when the operation timed out is not.

Reply:
Hi @ok2c ,

The 'actual' value is _not_ the number of milliseconds that have elapsed since midnight, January 1, 1970. It is a value >= deadline relative to the deadline because 'msecs' is equal to the timeout value in millis. The 'actual' value is really 'msecs' plus how much over the deadline we waited represented by 'waitTime'. This is useful because if the deadline is 30 seconds, but we ended up waiting 5 minutes, then we or the user can look for what I claim would be a bug, depending on how the Future is used.

Gary

Reply:
@garydgregory This is clearly not how I interpret this piece of code.
```              
 final long now = System.currentTimeMillis();
 if (now > deadline) {
     leaseRequest.failed(TimeoutValueException.ofMillis(deadline, now));
```

Reply:
Yes, for that call site, you are correct, my implementation is incorrect, I will adjust...

Reply:
@ok2c I've updated this PR to fix the above issue and also refactored our use of a deadline long and its redundant code into a new Deadline class.